src/registrar/views/domain.py

```diff
@@ -11,3 +17,60 @@ class DomainView(DomainPermission, DetailView):
     model = Domain
     template_name = "domain_detail.html"
     context_object_name = "domain"
+
+
+class DomainUsersView(DomainPermission, DetailView):
+    model = Domain
+    template_name = "domain_users.html"
+    context_object_name = "domain"
+
+
+class DomainAddUserForm(DomainPermission, forms.Form):
+
+    """Form for adding a user to a domain."""
+
+    email = forms.EmailField(label="Email")
+
+    def clean_email(self):
+        requested_email = self.cleaned_data["email"]
+        try:
+            User.objects.get(email=requested_email)
+        except User.DoesNotExist:
+            # TODO: send an invitation email to a non-existent user
+            raise forms.ValidationError("That user does not exist in this system.")
+        return requested_email
+
+
+class DomainAddUserView(DomainPermission, FormMixin, DetailView):
+    template_name = "domain_add_user.html"
+    model = Domain
+    form_class = DomainAddUserForm
+
+    def get_success_url(self):
+        return reverse("domain-users", kwargs={"pk": self.object.pk})
+
+    def post(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        form = self.get_form()
+        if form.is_valid():
+            return self.form_valid(form)
+        else:
+            return self.form_invalid(form)
+
+    def form_valid(self, form):
+        """Add the specified user on this domain."""
+        requested_email = form.cleaned_data["email"]
+        # look up a user with that email
+        # they should exist because we checked in clean_email
+        requested_user = User.objects.get(email=requested_email)
```

Comment:
Personally philosophy: if it is worth writing a note to explain how an error _couldn't_ occur, it is worth writing a try/catch block for that error. 👼

As a more practical note, I think once the email sending is implemented, that clean function can maybe be deleted, since the email will be sent from here, once the form is otherwise fully valid and submitted.

Reply:
This is about to change in follow-on work, but I agree with the sentiment and it will show up a couple times in that follow-on work.